python-client/giskard/models/base/__init__.py

```diff
@@ -244,30 +261,40 @@ def predict(self, dataset: Dataset) -> ModelPredictionResults:
               The `all_predictions` field will contain the predicted probabilities for all class labels for each example in the input dataset.
         """
         timer = Timer()
-        df = self.prepare_dataframe(dataset)
 
-        raw_prediction = self.predict_df(df)
+        # Read cache
+        cached_predictions = self.model_cache.read_from_cache(dataset.df[GISKARD_HASH_COLUMN])
+        missing = pd.isna(cached_predictions)
+        if len(missing.shape) > 1:
+            missing = np.any(missing, axis=1)
+
+        df = self.prepare_dataframe(dataset.slice(lambda x: dataset.df[missing], row_level=False))
+
+        if len(df) > 0:
+            raw_prediction = self.predict_df(df)
+            self.model_cache.set_cache(dataset.df[missing][GISKARD_HASH_COLUMN], raw_prediction)
+            cached_predictions[missing] = raw_prediction
 
```

Comment:
Let’s make sure that the naming here goes back to `raw_predictions`, the code after this should not be aware of the cache and should behave as usual (before your modifications).